lib/Model/DlibCnnHogModel/DlibCnnHogModel.php

```diff
@@ -0,0 +1,292 @@
+			->setParameter('id', $this->getId());
+		$resultStatement = $query->execute();
+		$data = $resultStatement->fetch(\PDO::FETCH_NUM);
+		$resultStatement->closeCursor();
+
+		if ((int)$data[0] <= 0) {
+			$query = $this->connection->getQueryBuilder();
+			$query->insert('facerecog_models')
+			->values([
+				'id' => $query->createNamedParameter($this->getId()),
+				'name' => $query->createNamedParameter($this->getName()),
+				'description' => $query->createNamedParameter($this->getDescription())
+			])
+			->execute();
+		}
+	}
+
+	public function open() {
+		$this->cfd = new \CnnFaceDetection($this->modelService->getFileModelPath($this->getId(), static::FACE_MODEL_FILES[self::I_MODEL_DETECTOR]));
+		$this->fld = new \FaceLandmarkDetection($this->modelService->getFileModelPath($this->getId(), static::FACE_MODEL_FILES[self::I_MODEL_PREDICTOR]));
+		$this->fr = new \FaceRecognition($this->modelService->getFileModelPath($this->getId(), static::FACE_MODEL_FILES[self::I_MODEL_RESNET]));
+	}
+
+	public function detectFaces(string $imagePath): array {
+		$detectedFaces = [];
+
+		$cnnFaces = $this->cfd->detect($imagePath, 0);
+		$hogFaces = dlib_face_detection($imagePath);
+
+		foreach ($cnnFaces as $proposedFace) {
+			$detectedFaces[] = $this->validateFace($proposedFace, $hogFaces);
+		}
+
+		return $detectedFaces;
+	}
+
+	public function detectLandmarks(string $imagePath, array $rect): array {
+		return $this->fld->detect($imagePath, $rect);
+	}
+
+	public function computeDescriptor(string $imagePath, array $landmarks): array {
+		return $this->fr->computeDescriptor($imagePath, $landmarks);
+	}
+
+	private function validateFace($proposedFace, $validateFaces) {
+		foreach ($validateFaces as $validateFace) {
+			$overlayPercent = $this->getOverlayPercent($proposedFace, $validateFace);
+			/**
+			 * The weak link in our default model is the landmark detector that
+			 * can't align profile faces correctly.
+			 * The Hog detector also fails and cannot detect these faces.
+			 *
+			 * So, if Hog detects it (Overlay > 80%), we know that the landmark
+			 * detector will do it too.
+			 * Just return it.
+			 */
+			if ($overlayPercent > 0.8) {
+				return $proposedFace;
+			}
+		}
+
+		/**
+		 * If Hog don't detect this face, they are probably in profile or rotated.
+		 * These are bad to compare, so we lower the confidence, to avoid clustering.
+		 */
+		$confidence = $proposedFace['detection_confidence'];
+		$proposedFace['detection_confidence'] = $confidence * 0.9;
+
+		return $proposedFace;
+	}
+
+	private function getOverlayPercent($rectP, $rectV): float {
+		// Proposed face rect
+		$leftP = $rectP['left'];
+		$rightP = $rectP['right'];
+		$topP = $rectP['top'];
+		$bottomP = $rectP['bottom'];
+
+		// Validate face rect
+		$leftV = $rectV['left'];
+		$rightV = $rectV['right'];
+		$topV = $rectV['top'];
+		$bottomV = $rectV['bottom'];
+
+		// If one rectangle is on left side of other
+		if ($leftP > $rightV || $leftV > $rightP)
```

Comment:
in all of these checks, add `>=` instead of `>`. I think you can get division with zero on line 289 if `$rightP==$leftP` etc.

Reply:
Ohh .. you're right..